File: mst/11733.cpp
/* prim - AC */
#include <iostream>
#include <queue>
#include <vector>
#include <utility>
#include <cstring>
using namespace std;
typedef pair<int, int> ii;
vector<pair<int, int> > graph[10001];
bool visited[10001];
int T, CA, N, M, A, mst, weight, src, dest, t;

int prim(int src){
	mst = CA; /* mst gasto interno de nodos conectados*/
	A++;
	priority_queue<ii, vector<ii>, greater<ii> > q;
	q.push(make_pair(0, src));
	while(!q.empty()){
		ii node = q.top(); q.pop();
		if(visited[node.second]) continue;
		visited[node.second] = true;

		if (node.first >= CA){ A++; mst += CA;}
		else mst += node.first;

		for(auto neighbour : graph[node.second]){
			if (!visited[neighbour.second]){				
				q.push(neighbour);
			}
		}
	}
	return mst;
}

void solution(){
	T = 0; /* total gastado */
	A = 0; /* aeropuertos por construir */
	memset(visited, false, sizeof visited);
	for(int i=1; i<=N; i++){
		if(!visited[i]){
			T += prim(i);
		}
	}
}

int main(){
	cin >> t;

	for(int cases=0; cases<t; cases++){
		cin >> N >> M >> CA;
		for(int i=0; i<M; i++){
			cin >> src >> dest >> weight;
			graph[src].push_back(make_pair(weight, dest));
			graph[dest].push_back(make_pair(weight, src));
		}

		solution();
		printf("Case #%d: %d %d\n", cases+1, T, A);
		for(int i=1; i<=N; i++) graph[i].clear();
	}

	return 0;
}

File: mst/11733.py
import sys, heapq
from queue import PriorityQueue as pq

def solution():
	global total
	total = 0
	for key, value in graph.items():
		if key not in visited:
			total += prim(key)
	return total

def prim(origen):
	global aeropuertos, visited
	mst = coste_aeropuerto
	aeropuertos += 1
	#pq = [(coste, destino) for coste, destino in graph[origen]]
	q = pq()
	for coste, destino in graph[origen]:
	 	q.put((coste, destino))
	#visited.add(origen)
	visited[origen] = True
	#heapq.heapify(pq)

	while not q.empty():
		#coste, destino = heapq.heappop(pq)
		coste, destino = q.get()
		if destino not in visited:
			#visited.add(destino)
			visited[destino] = True
			if coste >= coste_aeropuerto:
				mst += coste_aeropuerto
				aeropuertos += 1
			else:
				mst += coste
			for cost, next_ in graph[destino]:
				if next_ not in visited:
					#heapq.heappush(pq, (cost, next_))
					q.put((cost, next_))
				else: continue
		else: continue

	return mst


t = int(sys.stdin.readline())
for cases in range(t):
	ciudades, calles, coste_aeropuerto = [int(i) for i in sys.stdin.readline().split()]
	graph = {}
	for i in range(ciudades):
		graph.setdefault(i+1, [])
	
	for i in range(calles):
		origen, destino, coste = [int(i) for i in sys.stdin.readline().split()]
		
		graph[origen].append((coste, destino))
		graph[destino].append((coste, origen))
	aeropuertos, total = 0, 0
	#visited = set([])
	visited = {}
	total += solution()
	print('Case #%d: %d %d' % (cases+1, total, aeropuertos))
	

